pkg/admission/webhook.go

```diff
@@ -0,0 +1,112 @@
+package admission
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	herrors "github.com/horizoncd/horizon/core/errors"
+	"github.com/horizoncd/horizon/pkg/admission/models"
+	perror "github.com/horizoncd/horizon/pkg/errors"
+	"github.com/horizoncd/horizon/pkg/util/log"
+)
+
+const DefaultTimeout = 5 * time.Second
+
+var (
+	validatingWebhooks []Webhook
+)
+
+func Register(kind models.Kind, webhook Webhook) {
+	switch kind {
+	case models.KindValidating:
+		validatingWebhooks = append(validatingWebhooks, webhook)
+	}
+}
+
+func Clear() {
+	validatingWebhooks = nil
+}
+
+type validateResult struct {
+	err  error
+	resp *Response
+}
+
+type Request struct {
+	Operation    models.Operation       `json:"operation"`
+	Resource     string                 `json:"resource"`
+	ResourceName string                 `json:"resourceName"`
+	SubResource  string                 `json:"subResource"`
+	Version      string                 `json:"version"`
+	Object       interface{}            `json:"object"`
+	OldObject    interface{}            `json:"oldObject"`
+	Options      map[string]interface{} `json:"options,omitempty"`
+}
+
+type Response struct {
+	Allowed   *bool  `json:"allowed"`
+	Result    string `json:"result,omitempty"`
+	Patch     []byte `json:"patch,omitempty"`
+	PatchType string `json:"patchType,omitempty"`
```

Comment:
What are the optional values for PatchType?

Reply:
removed